File: 이정훈/9주차/[1106] 호텔.cpp
/*
* Author    : Lee JeongHun
* Date      : 2024.11.22(Fri)
* Runtime   : 0 ms
* Memory    : 2020 KB
* Algorithm : DP
*/

#include <iostream>
#include <vector>
#include <map>

using namespace std;

int main()
{
    ios_base::sync_with_stdio(0);
    cin.tie(0);
    cout.tie(0);

    // 입력
    int c, n;
    cin >> c >> n;

    map<int, int> table;
    for(int i = 0; i < n; ++i)
    {
        int cost, num;
        cin >> cost >> num;
        
        for(int j = 1; j <= c / num; ++j)
        {
            int key = num * j;
            int value = cost * j;

            if(table.find(key) == table.end() || table[key] > value)
            {
                table.insert(make_pair(key, value));
            }
        }
    }

    // 풀이
    vector<vector<int>> dp();

    return 0;
}

File: 강병호/9week/[1106] 호텔.py
'''
* Author    : BYEONGHO KANG
* Date      : 2024.11.29(Fri)
* Runtime   : 40 ms
* Memory    : 31120 KB
* Algorithm : 다이나믹 프로그래밍
'''

C, N = map(int, input().split())
cities = [list(map(int, input().split())) for _ in range(N)]

dp = [float('inf')] * (C + 101)
dp[0] = 0

for cost, customers in cities:
    for i in range(customers, C + 101):
        dp[i] = min(dp[i], dp[i - customers] + cost)

print(min(dp[C:]))
